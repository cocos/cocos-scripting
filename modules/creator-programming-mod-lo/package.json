{
  "name": "@cocos/creator-programming-mod-lo",
  "version": "1.10.0",
  "description": "",
  "main": "lib/mod-lo.js",
  "files": [
    "lib/**/*",
    "static/**/*.js",
    "static/**/*.mjs"
  ],
  "scripts": {
    "build:cjs-module-analyzer": "rollup -c ./scripts/build-cjs-module-analyzer.rollup.config.js",
    "build": "npm run build:cjs-module-analyzer & tsc",
    "clean": "tsc --build --clean",
    "prepublishOnly": "npm run build"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {

    "@babel/core": "7.21.3",
    "@babel/generator": "7.16.0",
    "@babel/helper-module-imports": "7.18.6",
    "@babel/parser": "7.22.4",
    "@babel/plugin-proposal-dynamic-import": "7.16.0",
    "@babel/plugin-syntax-top-level-await": "7.14.5",
    "@babel/plugin-transform-modules-systemjs": "7.16.0",
    "@babel/preset-env": "7.16.4",
    "@babel/template": "7.16.0",

    "@endo/cjs-module-analyzer": "0.2.11",
    "@rollup/plugin-node-resolve": "13.0.6",

    "@babel/plugin-transform-classes": "7.16.0",
    "@babel/code-frame": "7.21.4",
    "@babel/helper-compilation-targets": "7.20.7",
    "@babel/helper-module-transforms": "7.21.2",
    "@babel/helpers": "7.21.0",
    "@babel/traverse": "7.22.4",
    
    "@cocos/creator-programming-babel-preset-cc": "*",
    "@cocos/creator-programming-common": "*",
    "@cocos/creator-programming-import-maps": "*",
    "acorn": "8.6.0",
    "acorn-walk": "8.2.0",
    "cjs-module-lexer": "1.2.2",
    "dedent": "0.7.0",
    "fs-extra": "9.1.0",
    "globby": "11.0.4",
    "javascript-stringify": "2.1.0",
    "magic-string": "0.25.7",
    "minimatch": "9.0.0",
    "resolve": "1.20.0",
    "source-map": "0.7.3"
  },
  "devDependencies": {

    "@babel/helper-string-parser": "7.21.5",
    "@babel/helper-validator-identifier": "7.19.1",

    "@babel/helper-transform-fixture-test-runner": "7.16.0",
    "@babel/plugin-transform-modules-commonjs": "7.16.0",

    "@cocos/creator-programming-test-utils": "*",
    "@endo/cjs-module-analyzer": "^0.2.7",
    "@rollup/plugin-node-resolve": "13.0.6",
    "@types/babel__core": "7.1.20",
    "@babel/types": "7.22.4",
    "@types/babel__generator": "7.6.4",
    "@types/babel__parser": "7.1.1",
    "@types/babel__preset-env": "7.9.2",
    "@types/babel__template": "7.4.1",
    "@types/babel__traverse": "7.18.3",
    "@types/dedent": "0.7.0",
    "@types/fs-extra": "9.0.13",
    "@types/node": "16.11.10",
    "@types/resolve": "1.20.1",
    "rollup": "2.60.1",
    "typescript": "4.5.2"
  },
  "publishConfig": {
    "access": "public"
  },
  "gitHead": "a63a43d774f85815a9f13467161138b2041636e6"
}
