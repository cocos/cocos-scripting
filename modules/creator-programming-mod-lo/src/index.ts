export * from './mod-lo';
export * from './cjs/share';
