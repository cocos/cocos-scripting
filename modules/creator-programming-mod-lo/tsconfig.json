{
    "extends": "../../tsconfig.json",
    "compilerOptions": {
        "module": "CommonJS",
        "rootDir": "src",
        "outDir": "lib",
        "types": [
            "./@types/babel-helper-module-imports"
        ]
    },
    "exclude": [
        "lib",
        "test",
        "src/transformer/swc/**/*.ts"
    ]
}