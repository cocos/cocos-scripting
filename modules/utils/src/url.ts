import { URL } from 'url';

export function parseExtensionName(pathname: string) {
    const segments = pathname.split('/');
    if (segments.length > 0) {
        const last = segments[segments.length - 1];
        if (last.length > 1) {
            const iLastDot = last.lastIndexOf('.');
            if (iLastDot >= 0) {
                return last.substr(iLastDot);
            }
        }
    }
    return '';
}

export function hasExtension(pathname: string) {
    return !!parseExtensionName(pathname);
}

export function replaceExtension(url: URL, ext: string) {
    const parts = url.pathname.split('/');
    if (parts.length !== 0) {
        const baseName = parts[parts.length - 1];
        const dot = baseName.lastIndexOf('.');
        if (dot >= 0) {
            const originalExtension = baseName.substr(dot + 1);
            parts[parts.length - 1] = `${baseName.substr(0, dot)}${ext}`;
            assignPathname(url, parts.join('/'));
            return '.' + originalExtension;
        }
    }
}

export function tryParseURL(input: string, base?: URL) {
    try {
        return new URL(input, base);
    } catch {
        return;
    }
}

export function assignPathname(url: URL, pathname: string) {
    if (!url.pathname.startsWith('/')) {
        console.warn(`Cannot change pathname of URL ${url} since it's a 'cannot-be-a-base-URL' URL.`);
    }
    // https://gist.github.com/shrinktofit/50f2380782b6219ed6a4cb3a210a75f7
    url.pathname = pathname;
    return url;
}

export function getBaseName(url: URL) {
    const segments = url.pathname.split('/');
    if (segments.length === 0) {
        return '';
    } else {
        const last = segments[segments.length - 1];
        const iDot = last.lastIndexOf('.');
        if (iDot >= 0) {
            return last.substr(0, iDot);
        } else {
            return last;
        }
    }
}

export function hasFileProtocol(url: URL) {
    return url.protocol === 'file:';
}
