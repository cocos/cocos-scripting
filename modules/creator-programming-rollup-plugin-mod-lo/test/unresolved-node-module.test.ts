
import * as rollup from 'rollup';
import { ModLo } from '@cocos/creator-programming-mod-lo';
import rpModLo from '../src/index';
import ps from 'path';
import { pathToFileURL } from 'url';

test('rollup-plugin-mod-lo unresolved node module', async () => {
    const modLo = new ModLo({
        _compressUUID: (): string => '',
        _helperModule: rpModLo.helperModule,
    });

    const bundle = await rollup.rollup({
        input: pathToFileURL(ps.join(__dirname, 'inputs', 'unresolved-node-module', 'index.cjs')).href,
        plugins: [
            rpModLo({
                modLo,
            }),
        ],
    });

    const { output } = await bundle.generate({
        format: 'systemjs',
        exports: 'auto',
        preserveModules: true,
    });

    for (const chunk of output) {
        if (chunk.type !== 'chunk') {
            continue;
        }
        expect(chunk.code).toMatchSnapshot(`chunk: ${chunk.fileName}`);
    }
});