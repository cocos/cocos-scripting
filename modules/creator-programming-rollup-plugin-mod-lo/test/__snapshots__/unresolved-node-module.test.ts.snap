// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rollup-plugin-mod-lo unresolved node module: chunk: _virtual/cjs-loader.mjs.js 1`] = `
"System.register(['./rollupPluginModLoBabelHelpers.js'], (function (exports) {
  'use strict';
  var _createClass, _classCallCheck;
  return {
    setters: [function (module) {
      _createClass = module.createClass;
      _classCallCheck = module.classCallCheck;
    }],
    execute: (function () {

      var CjsLoader = /*#__PURE__*/function () {
        function CjsLoader() {
          _classCallCheck(this, CjsLoader);
          this._registry = {};
          this._moduleCache = {};
        }

        /**
         * Defines a CommonJS module.
         * @param id Module ID.
         * @param factory The factory.
         * @param resolveMap An object or a function returning object which records the module specifier resolve result.
         * The later is called as "deferred resolve map" and would be invocated right before CommonJS code execution.
         */
        _createClass(CjsLoader, [{
          key: "define",
          value: function define(id, factory, resolveMap) {
            this._registry[id] = {
              factory: factory,
              resolveMap: resolveMap
            };
          }

          /**
           * Requires a CommonJS module.
           * @param id Module ID.
           * @returns The module's \`module.exports\`.
           */
        }, {
          key: "require",
          value: function require(id) {
            return this._require(id);
          }
        }, {
          key: "throwInvalidWrapper",
          value: function throwInvalidWrapper(requestTarget, from) {
            throw new Error("Module '".concat(requestTarget, "' imported from '").concat(from, "' is expected be an ESM-wrapped CommonJS module but it doesn't."));
          }
        }, {
          key: "_require",
          value: function _require(id, parent) {
            var cachedModule = this._moduleCache[id];
            if (cachedModule) {
              return cachedModule.exports;
            }
            var module = {
              id: id,
              exports: {}
            };
            this._moduleCache[id] = module;
            this._tryModuleLoad(module, id);
            return module.exports;
          }
        }, {
          key: "_resolve",
          value: function _resolve(specifier, parent) {
            return this._resolveFromInfos(specifier, parent) || this._throwUnresolved(specifier, parent);
          }
        }, {
          key: "_resolveFromInfos",
          value: function _resolveFromInfos(specifier, parent) {
            var _cjsInfos$parent$reso, _cjsInfos$parent;
            if (specifier in cjsInfos) {
              return specifier;
            }
            if (!parent) {
              return;
            }
            return (_cjsInfos$parent$reso = (_cjsInfos$parent = cjsInfos[parent]) === null || _cjsInfos$parent === void 0 ? void 0 : _cjsInfos$parent.resolveCache[specifier]) !== null && _cjsInfos$parent$reso !== void 0 ? _cjsInfos$parent$reso : undefined;
          }
        }, {
          key: "_tryModuleLoad",
          value: function _tryModuleLoad(module, id) {
            var threw = true;
            try {
              this._load(module, id);
              threw = false;
            } finally {
              if (threw) {
                delete this._moduleCache[id];
              }
            }
          }
        }, {
          key: "_load",
          value: function _load(module, id) {
            var _this$_loadWrapper = this._loadWrapper(id),
              factory = _this$_loadWrapper.factory,
              resolveMap = _this$_loadWrapper.resolveMap;
            var vendorRequire = this._createRequire(module);
            var require = resolveMap ? this._createRequireWithResolveMap(typeof resolveMap === 'function' ? resolveMap() : resolveMap, vendorRequire) : vendorRequire;
            factory(module.exports, require, module);
          }
        }, {
          key: "_loadWrapper",
          value: function _loadWrapper(id) {
            if (id in this._registry) {
              return this._registry[id];
            } else {
              return this._loadHostProvidedModules(id);
            }
          }
        }, {
          key: "_loadHostProvidedModules",
          value: function _loadHostProvidedModules(id) {
            return {
              factory: function factory(_exports, _require, module) {
                if (typeof require === 'undefined') {
                  throw new Error("Current environment does not provide a require() for requiring '".concat(id, "'."));
                }
                try {
                  module.exports = require(id);
                } catch (err) {
                  throw new Error("Exception thrown when calling host defined require('".concat(id, "')."), {
                    cause: err
                  });
                }
              }
            };
          }
        }, {
          key: "_createRequire",
          value: function _createRequire(module) {
            var _this = this;
            return function (specifier) {
              return _this._require(specifier, module);
            };
          }
        }, {
          key: "_createRequireWithResolveMap",
          value: function _createRequireWithResolveMap(requireMap, originalRequire) {
            return function (specifier) {
              var resolved = requireMap[specifier];
              if (resolved) {
                return originalRequire(resolved);
              } else {
                throw new Error('Unresolved specifier ' + specifier);
              }
            };
          }
        }, {
          key: "_throwUnresolved",
          value: function _throwUnresolved(specifier, parentUrl) {
            throw new Error("Unable to resolve ".concat(specifier, " from ").concat(parent, "."));
          }
        }]);
        return CjsLoader;
      }();
      var loader = exports('default', new CjsLoader());

    })
  };
}));
"
`;

exports[`rollup-plugin-mod-lo unresolved node module: chunk: _virtual/index.cjs.js 1`] = `
"System.register(['./cjs-loader.mjs.js', './javascript_%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20export%20const%20__cjsMetaURL%20%3D%20\\'fs\\'%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20.js'], (function (exports, module) {
  'use strict';
  var loader, __cjsMetaURL$1;
  return {
    setters: [function (module) {
      loader = module["default"];
    }, function (module) {
      __cjsMetaURL$1 = module.__cjsMetaURL;
    }],
    execute: (function () {

      var _cjsExports; exports('default', _cjsExports);
      var __cjsMetaURL = exports('__cjsMetaURL', module.meta.url);
      loader.define(__cjsMetaURL, function (exports$1, require, module, __filename, __dirname) {
        // #region ORIGINAL CODE

        require('fs');

        // #endregion ORIGINAL CODE

        exports('default', _cjsExports = module.exports);
      }, function () {
        return {
          'fs': __cjsMetaURL$1
        };
      });

    })
  };
}));
"
`;

exports[`rollup-plugin-mod-lo unresolved node module: chunk: _virtual/index.mjs.js 1`] = `
"System.register(['./index.cjs.js', './cjs-loader.mjs.js'], (function (exports, module) {
  'use strict';
  var __cjsMetaURL, loader;
  return {
    setters: [function (module) {
      __cjsMetaURL = module.__cjsMetaURL;
      exports({ __cjsMetaURL: module.__cjsMetaURL, 'default': module["default"] });
    }, function (module) {
      loader = module["default"];
    }],
    execute: (function () {

      // I am the facade module who provides access to the CommonJS module './index.cjs'~
      if (!__cjsMetaURL) {
        loader.throwInvalidWrapper('./index.cjs', module.meta.url);
      }
      loader.require(__cjsMetaURL);

    })
  };
}));
"
`;

exports[`rollup-plugin-mod-lo unresolved node module: chunk: _virtual/javascript_%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20export%20const%20__cjsMetaURL%20%3D%20'fs'%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20.js 1`] = `
"System.register([], (function (exports) {
	'use strict';
	return {
		execute: (function () {

			var __cjsMetaURL = exports('__cjsMetaURL', 'fs');

		})
	};
}));
"
`;

exports[`rollup-plugin-mod-lo unresolved node module: chunk: _virtual/rollupPluginModLoBabelHelpers.js 1`] = `
"System.register([], (function (exports) {
  'use strict';
  return {
    execute: (function () {

      exports({
        classCallCheck: _classCallCheck,
        createClass: _createClass
      });

      function _classCallCheck(instance, Constructor) {
        if (!(instance instanceof Constructor)) {
          throw new TypeError("Cannot call a class as a function");
        }
      }
      function _defineProperties(target, props) {
        for (var i = 0; i < props.length; i++) {
          var descriptor = props[i];
          descriptor.enumerable = descriptor.enumerable || false;
          descriptor.configurable = true;
          if ("value" in descriptor) descriptor.writable = true;
          Object.defineProperty(target, descriptor.key, descriptor);
        }
      }
      function _createClass(Constructor, protoProps, staticProps) {
        if (protoProps) _defineProperties(Constructor.prototype, protoProps);
        if (staticProps) _defineProperties(Constructor, staticProps);
        Object.defineProperty(Constructor, "prototype", {
          writable: false
        });
        return Constructor;
      }

    })
  };
}));
"
`;
